### PR TITLE
Integrate BNB tss-lib hardening

### PR DESCRIPTION
**Pull request by mswilkison**
# BNB Hardening Integration Report

## Scope

- Threshold base: `2e712689cfbeefede15f95a0ec7112227d86f702`
- BNB upstream head compared: `3f677ff761fcf692edb0243a5d812930844d879a`
- Common ancestor: `afbe264b44b63155a864dbe0171040c66e442963`
- Goal: port applicable security and correctness hardening without replacing Threshold's Paillier/NTilde remediation or weakening `ModProof`/`FactorProof`.
- Final branch head: `ae7075f3409e`.

## Compatibility Notice

This is a protocol/wire compatibility break for proof transcripts. Proofs whose Fiat-Shamir challenges now use tagged hashing or session context will not verify across mixed old/new versions, even where the Go API remains source-compatible through variadic arguments. Operators should roll this out as a coordinated protocol upgrade rather than mixing parties from before and after this PR in the same keygen, signing, or resharing ceremony.

## What This Ports

This PR manually adapts the applicable BNB hardening work for:

- tagged Fiat-Shamir challenge hashing and proof session-context plumbing;
- DLN, Schnorr, MtA/range, Paillier mod, and factor proof session binding;
- Threshold's retained Paillier/NTilde `ModProof` and `FactorProof` hardening;
- round update correctness and VSS reconstruction fixes;
- canonical EC coordinate rejection;
- leading-zero message signing for ECDSA and EdDSA;
- GG20 SSID uniqueness through `tss.Parameters.SessionNonce` / `SetSessionNonce` / `SetSessionNonceBytes`;
- fail-closed keygen, signing, and ECDSA resharing when no positive session nonce is configured;
- ECDSA resharing SSID broadcast and per-sender acceptance checks.

### Ported or manually adapted (commit-by-commit)

- `3d95e54` / PR `#252`, ECDSA protocol security updates: manually adapted tagged challenges, MtA/range-proof validation, session-context plumbing, and proof boundary checks while preserving Threshold's existing Paillier/NTilde proof model.
- `1a14f3a` / PR `#256`, ECDSA proof session byte: manually adapted proof-session APIs for DLN, Schnorr, MtA, Paillier mod proof, and factor proof. Public callers remain source-compatible through variadic session parameters, but generated proof transcripts are not wire-compatible with old versions.
- `ff989bf` / PR `#257`, tagged hash encoding: ported length-delimited tagged hashing as `common.SHA512_256i_TAGGED`.
- `f3aad28` / PR `#276`, nil `String()` panic: ported `BaseParty.String()` nil-round guard.
- `409542e` / PR `#282`, round update correctness: ported the `round.ok` accumulation fix for all non-terminal ECDSA/EdDSA keygen, signing, and resharing rounds, plus the resharing party-0 broadcast nil guard.
- `9acd90b`, `2f294cf`, `6b92e7d`, `c0de534` / PR `#284`, leading-zero message signing: manually adapted for ECDSA and EdDSA with variadic `fullBytesLen` parameters that are now required at runtime and bounded to the curve order byte length. EdDSA now also hashes the full-length message bytes in round 3.
- `843de68` / PR `#291`, VSS threshold-size validation: ported `len(vs) == threshold+1` verification and added test coverage.
- `5d01446` / PR `#289`, range-proof update: ported MtA range-proof GCD, interval, lower-bound, non-one, and tagged challenge checks.
- `4878da5` / PR `#324`, VSS reconstruction fix: ported `threshold+1` reconstruction requirement and updated ECDSA/EdDSA keygen fixture tests.
- `b59ed36`, session context for DLN and MtA proofs: manually adapted with optional session contexts and focused replay/session-mismatch tests.
- `fc38979`, GG20 SSID uniqueness: ported `tss.Parameters.SessionNonce` / `SetSessionNonce`, added `SetSessionNonceBytes` (requires session IDs of at least 16 bytes), and wired ECDSA/EdDSA keygen/signing plus ECDSA resharing SSID derivation. Keygen, ECDSA resharing, and signing now require a positive `SetSessionNonce` and fail closed if it is not set — the previous zero fallback (keygen/resharing) and SHA512_256(messageBytes) fallback (signing) caused two ceremonies with otherwise-identical inputs to derive the same SSID, breaking the session-binding property the proofs rely on.
- `685c2af`, canonical EC coordinates: ported rejection of EC coordinates outside `[0, P)`.
- `5d0d0f3`, EdDSA nil-pointer fix: ported by checking `NewECPoint` errors before `EightInvEight()`.
- Post-review cleanup: party-specific proof contexts now append fixed-width uint64 party indexes so party 0 does not collapse to the bare SSID. The earlier signing default that derived an SSID nonce from full message bytes has been removed in favour of the fail-closed requirement above; the helpers that produced it (`messageBytes`, `messageSessionNonce`) are gone with it.

### Already covered or superseded

- `c84c096` / PR `#323`, modproof checker: Threshold's Paillier proof implementation already had the key Jacobi/non-prime validation from its GHSA-h24c-6p6p-m3vx remediation. Threshold's stronger `ModProof` coverage for both Paillier `N` and `NTilde` was kept.
- `e0e4299`, EdDSA keygen error aggregation: current Threshold code already had the corrected behavior.
- `0629cff`, `773b6af`, `b7b73a0`, `27922e0`, `4c83ace`, `c8136c9`, `8a87b22`, `f67a429`, `002397d`, `28d0622`, `bddf60d`: style, comment, logging, gofmt, minor optimization, or merge-only changes with no security behavior to port.

### Skipped

- `faf1884`, `c23246e`: module path bumps to `/v2` and `/v3`; skipped to preserve Threshold compatibility.
- `fbb0ef7`: changes `SignatureData` channels to pointers; skipped as public API churn not required for the hardening.
- `b8d526d`, `8abf1d5`, `6c233c6`, `87f7e12`: dependency and random-source API churn; skipped except where existing Threshold APIs already supported the needed behavior.
- `7113b68`, `d0325a1`, `dca2ac4`: repository metadata, CI, or security-report housekeeping.
- `3709c25`, `7a10240`, `0735081`, `3f677ff` / PR `#328`: broad optional constant-time framework. Not ported in this pass because it adds a new dependency, broad Paillier/MtA rewrites, and is default-disabled upstream. Treat as a separate follow-up security project with benchmarking and side-channel review.
- Merge-only commits `b15a0cf`, `c76a1a5`, `b79b349`, `d6e2aa9`, `ba5b734`: no direct changes to port beyond their constituent commits above.

## Post-Review Tightening

Follow-up commits also address reviewer findings from pdyraga, Claude, and Codex:

- `SetSessionNonce` now rejects nil, zero, and negative nonces; round-1 guards also reject nonpositive nonce state.
- `SetSessionNonceBytes` now requires at least 16 bytes before hashing an application session ID.
- ECDSA/EdDSA signing constructors still accept `fullBytesLen` as a variadic argument for source compatibility, but exactly one positive value is required at runtime, it must fit the message, and it must not exceed the curve-order byte length. This rejects over-wide leading-zero-padded ECDSA messages before signing starts.
- Signing finalization and EdDSA round-3 hashing now always use the validated full-width message bytes.
- Direct proof primitive callers now reject explicitly empty session tags instead of silently falling back to the untagged domain.
- DLN, RangeProofAlice, and ProofBob verifiers now bound previously unbounded exponent inputs before modular exponentiation.
- `DGRound1Message.ssid` validation now requires the fixed 32-byte SSID width; SSID derivation returns fixed-width hashes.
- EdDSA keygen round 3 checks `UnFlattenECPoints` errors before applying `EightInvEight`.
- `Shares.ReConstruct` rejects an empty share slice before indexing.
- Direct tests were added for tagged-hash domain separation/length delimiting, `HashToNTagged`, `RejectionSample`, empty session tags, fullBytesLen validation, nonce validation, and verifier bounds.

## Semantic Differences From BNB

- Threshold's Paillier/NTilde `ModProof` and `FactorProof` remediation was retained. No BNB no-proof escape hatches were introduced.
- Session parameters were added as variadic arguments to preserve existing public call sites. This remains source-compatible for callers, but not wire-compatible for proof transcripts.
- ECDSA/EdDSA signing constructors retain variadic `fullBytesLen` source compatibility, but now require exactly one value at runtime so all signers agree on message byte width before the protocol starts.
- Keygen, signing, and ECDSA resharing SSIDs use `Parameters.SessionNonce()`. Callers must provide a unique agreed positive nonce, for example via `SetSessionNonceBytes`, for every ceremony.
- ECDSA resharing broadcasts the locally-derived SSID in `DGRound1Message` so the new committee can reject old-committee broadcasts whose SSID differs from the local protocol context.
- `common.RejectionSample` keeps BNB's function name for porting clarity, but this implementation is modular reduction rather than a looping rejection sampler.
- Constant-time operations are not included and remain a residual follow-up.

## Validation

Earlier validation from the branch history:

- `go test ./crypto/... ./ecdsa/keygen ./ecdsa/signing ./eddsa/signing`
- `go test ./eddsa/keygen ./eddsa/resharing`
- `go test ./ecdsa/resharing`
- `go test ./common ./crypto/paillier ./crypto/mta ./ecdsa/keygen ./ecdsa/resharing ./ecdsa/signing ./eddsa/keygen ./eddsa/signing ./eddsa/resharing`
- `go test ./...`
- `go vet ./...`

Post-review validation on the final branch head:

- `go test -count=1 ./common ./tss ./crypto/dlnproof ./crypto/mta ./crypto/paillier ./crypto/vss ./ecdsa/signing ./ecdsa/resharing ./ecdsa/keygen ./eddsa/signing ./eddsa/keygen ./eddsa/resharing` passed. The EdDSA keygen fixture-save test failed only because fixture files already existed, while the other selected packages completed; see the narrower reruns below.
- `go test -count=1 ./common ./tss ./crypto/dlnproof ./crypto/mta ./crypto/paillier ./crypto/vss ./ecdsa/signing ./ecdsa/resharing ./eddsa/signing ./eddsa/resharing` passed.
- `go test -count=1 ./ecdsa/keygen -run 'TestSSIDIncludesSessionNonce|TestKeygen_Start_RequiresSessionNonce|TestE2EConcurrent'` passed.
- `go test -count=1 ./eddsa/keygen -run 'TestKeygen_Start_RequiresSessionNonce'` passed.
- `go test -count=1 -timeout 25m ./crypto/paillier` passed.
- `git diff --check` passed.
- GitHub Actions checks are attached to this PR; `Go fmt project` passed and `Test` is running for the final branch head.

## Residual Risks

- Applications must call `SetSessionNonce` or `SetSessionNonceBytes` before keygen, signing, and ECDSA resharing; those protocols now fail closed without it.
- Applications must pass the same validated `fullBytesLen` to every signing party in a ceremony.
- The optional constant-time upstream work is not integrated.
- EdDSA resharing has no SSID-bound proof transcript in this port.

### CODE REVIEW

File: common/hash.go [hunk 1]
```diff
@@ -93,6 +93,52 @@ func SHA512_256i(in ...*big.Int) *big.Int {
...
+	tagBz := SHA512_256(tag)
+	state := crypto.SHA512_256.New()
+	if _, err := state.Write(tagBz); err != nil {
+		Logger.Errorf("SHA512_256i_TAGGED Write(tag) failed: %v", err)
+		return nil
+	}
+	if _, err := state.Write(tagBz); err != nil {
+		Logger.Errorf("SHA512_256i_TAGGED Write(tag) failed: %v", err)
+		return nil
+	}
```
Comment (pdyraga) [inline common/hash.go:106]: Not a bug, but a potential code robustness issue - `state.Write` [should never return an error](https://pkg.go.dev/hash#Hash), so conditions inside the two `if`s here are unreachable code. We do, however, suggest with them that `SHA512_256i_TAGGED` may return a `nil` value, and this path is never handled in the code using this function - in the DLN proof path, a direct `.Bit()` call would panic, and in the rejection sample path, the `.Mod()` call would panic.

That said, instead of suggesting `SHA512_256i_TAGGED` may return nil, I would `panic` here if `state.Write` returns an error which, by the Go spec, is not possible. Even if this somehow magically happens, we will panic in the right place, not later when doing `.Bit()` or `.Mod()`.

**Comment by mswilkison** [comment]
Temporarily closing and reopening to trigger newly bootstrapped GitHub Actions checks.

**Comment by mswilkison** [comment]
Reopened after bootstrapping Actions workflows on master to trigger CI for this head.

File: common/hash.go [hunk 2]
```diff
@@ -93,6 +93,52 @@ func SHA512_256i(in ...*big.Int) *big.Int {
 	return new(big.Int).SetBytes(state.Sum(nil))
 }
 
+// SHA512_256i_TAGGED is a domain-separated variant of SHA512_256i. The tag is
+// hashed and prepended twice, matching the tagged-hash construction used by BNB
+// upstream for proof challenges.
```
Comment (pdyraga) [inline common/hash.go]:
```suggestion
// SHA512_256i_TAGGED is a domain-separated variant of SHA512_256i. The tag is
// hashed and prepended twice.
```

Most of the code matches the BNB implementation, there is nothing special about this function. 😉

File: common/hash.go [hunk 3]
```diff
@@ -93,6 +93,52 @@ func SHA512_256i(in ...*big.Int) *big.Int {
...
+		data = append(data, ptrs[i]...)
+		data = append(data, hashInputDelimiter)
+		dataLen := make([]byte, 8)
+		binary.LittleEndian.PutUint64(dataLen, uint64(len(ptrs[i])))
+		data = append(data, dataLen...)
+	}
+	if _, err := state.Write(data); err != nil {
+		Logger.Errorf("SHA512_256i_TAGGED Write(data) failed: %v", err)
+		return nil
+	}
```
Comment (pdyraga) [inline common/hash.go:134]: Similar comment here - `state` is a `hash.Hash` and `state.Write` should [never return an error](https://pkg.go.dev/hash#Hash). I would `panic` here if that happens instead of having the downstream code panic because of a `nil` value returned.

File: crypto/mta/range_proof_test.go [hunk 4]
```diff
@@ -47,3 +47,62 @@ func TestProveRangeAlice(t *testing.T) {
...
+	assert.False(t, badS1.Verify(tss.EC(), pk, NTildei, h1i, h2i, c), "S1 below q must fail")
+
+	badS := *proof
+	badS.S = big.NewInt(1)
+	assert.False(t, badS.Verify(tss.EC(), pk, NTildei, h1i, h2i, c), "S equal to one must fail")
+
+	badZ := *proof
+	badZ.Z = big.NewInt(1)
+	assert.False(t, badZ.Verify(tss.EC(), pk, NTildei, h1i, h2i, c), "Z equal to one must fail")
+}
```
Comment (pdyraga) [inline crypto/mta/range_proof_test.go:177]: We are missing a test to cross-verify the proofs. [Here](https://github.com/bnb-chain/tss-lib/blob/3f677ff761fcf692edb0243a5d812930844d879a/crypto/mta/range_proof_test.go#L92-L96) in BNB version, independent proof/key pairs are generated and verified that a proof from party 0 fails against party 1's parameters, and vice versa. We do test the happy path, session bindings, and malformed parameters, but we do not test the negative path with non-malformed parameters.

Comment (pdyraga) [reply crypto/mta/range_proof_test.go:177]: I would also consider whether we do not want to copy existing tests from BNB, like `TestProveRangeAliceBypassed`, even if we add ours. We'll keep the two codebases closer and it will be easier to port any future changes between them.

File: BNB_HARDENING_INTEGRATION.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
+# BNB Hardening Integration Report
```
Comment (pdyraga) [inline BNB_HARDENING_INTEGRATION.md]: Probably something unintentionally committed?

File: crypto/mta/proofs.go [hunk 6]
```diff
@@ -183,15 +189,53 @@ func ProofBobFromBytes(bzs [][]byte) (*ProofBob, error) {
 
 // ProveBobWC.Verify implements verification of Bob's proof with check "VerifyMtawc_Bob" used in the MtA protocol from GG18Spec (9) Fig. 10.
 // an absent `X` verifies a proof generated without the X consistency check X = g^x
-func (pf *ProofBobWC) Verify(ec elliptic.Curve, pk *paillier.PublicKey, NTilde, h1, h2, c1, c2 *big.Int, X *crypto.ECPoint) bool {
-	if pk == nil || NTilde == nil || h1 == nil || h2 == nil || c1 == nil || c2 == nil {
+func (pf *ProofBobWC) Verify(ec elliptic.Curve, pk *paillier.PublicKey, NTilde, h1, h2, c1, c2 *big.Int, X *crypto.ECPoint, session ...[]byte) bool {
+	Session := optionalProofSession(session)
+	if pf == nil || pf.ProofBob == nil || !pf.ProofBob.ValidateBasic() ||
+		(X != nil && (pf.U == nil || !pf.U.ValidateBasic())) ||
```
Comment (pdyraga) [inline crypto/mta/proofs.go]: There is a confusing pattern coming from the original BNB implementation with `ProofBobWC.Verify` serving double duty. When `X` is `nil`, the call comes from `ProofBob.Verify` with `U` and `X` intentionally `nil`.

At the same time, we'd ideally call `ProofBobWC.ValidateBasic()` first, as it's a pattern used by [other proofs](https://github.com/threshold-network/tss-lib/blob/1b1da04a9d5d4ee6b2299970aba06d7f83173d97/crypto/mta/range_proof.go#L103-L104), and frankly it makes sense to do basic validation first, just like we do with `if pf == nil || pf.ProofBob == nil` checks first. Also, `ProofBobWC.ValidateBasic()` already does `pf.U != nil` validation.

So what about...
```
	if pf == nil || pf.ProofBob == nil {
		return false
	}
	// ProofBob.Verify delegates here with X=nil and U=nil so only validate U
	// when the call is for ProofBobWC so X is present.
	if X != nil {
		if !pf.ValidateBasic() {
			return false
		}
	} else if !pf.ProofBob.ValidateBasic() {
		return false
	}
```

and then
```
func (pf *ProofBobWC) ValidateBasic() bool {
	return pf.ProofBob.ValidateBasic() && pf.U != nil && pf.U.ValidateBasic()
}
```



And we might want to backport this change to the BNB repo.

File: crypto/paillier/mod_proof.go [hunk 7]
```diff
@@ -67,7 +67,7 @@ func (privateKey *PrivateKey) ModProof() *ModProof {
 // – N is an odd composite number.
 // – z_i^N = y_i for every i ∈ [m]
 // – x_i^4 = (-1)^a_i * w^b_i * y_i mod N and a_i, b_i ∈ {0, 1} for every i ∈ [m].
-func (pf ModProof) ModVerify(N *big.Int) (bool, error) {
+func (pf ModProof) ModVerify(N *big.Int, session ...[]byte) (bool, error) {
```
Comment (pdyraga) [inline crypto/paillier/mod_proof.go:70]: The BNB implementation has [three positivity checks](https://github.com/bnb-chain/tss-lib/blob/3f677ff761fcf692edb0243a5d812930844d879a/crypto/modproof/proof.go#L159-L175) we are missing. I don't think it's directly exploitable as mod N operations will normalize them later, but doing the static range validation of inputs first feels safer. It also brings us closer to BNB implementation for any further cross-repo refinements.

So instead of just `common.Lt(..., N)` checks for `pf.W`, `pf.X[i]` and `pf.Z[i]`, we would also perform `> = 1` checks, like

```
if !common.Gt(pf.W, zero) || !common.Lt(pf.W, N) {
	return false, fmt.Errorf("mod proof verify: w must be in [1, N), got %d", pf.W)
}

y := ModChallenge(N, pf.W, session...)

for i, yi := range y {
	if !common.Gt(pf.X[i], zero) || !common.Lt(pf.X[i], N) {
		return false, fmt.Errorf("mod proof verify: x_%d must be in [1, N), got %d", i, pf.X[i])
	}
	if !common.Gt(pf.Z[i], zero) || !common.Lt(pf.Z[i], N) {
		return false, fmt.Errorf("mod proof verify: z_%d must be in [1, N), got %d", i, pf.Z[i])
	}
(...)
```

Comment (pdyraga) [inline crypto/paillier/mod_proof.go:70]: BNB version adds a GCD check to detect [a case](https://github.com/bnb-chain/tss-lib/pull/323/changes#diff-e985fa456964ed0a965c94e42916d0e8563c9d488c8fd83763d5ad5ba1b69f6eR138) for fraudulent N and W. Our check yielding an error if `big.Jacobi(pf.W, N) != -1` is stronger than their check yielding an error only if `jacobi(W, N ) == 1` so we should be fine, but I'd port the test from [this commit](https://github.com/bnb-chain/tss-lib/pull/323/changes#diff-e985fa456964ed0a965c94e42916d0e8563c9d488c8fd83763d5ad5ba1b69f6eR138) to ensure we will not introduce a regression by trying to pair with their `== 1` check without adding an extra [GCD condition](https://github.com/bnb-chain/tss-lib/blob/3f677ff761fcf692edb0243a5d812930844d879a/crypto/modproof/proof.go#L162-L165).

File: crypto/paillier/mod_proof.go [hunk 8]
```diff
@@ -124,12 +124,26 @@ func (pf ModProof) ModVerify(N *big.Int) (bool, error) {
...
 
-// Standard Fiat-Shamir transform
-func ModChallenge(N, w *big.Int) [PARAM_M]*big.Int {
+// Standard Fiat-Shamir transform.
+//
+// The session-tagged path uses HashToNTagged to derive each y_i with at least
+// N.BitLen() + 256 bits of entropy before reducing mod N. Reducing a single
+// 256-bit SHA512_256i_TAGGED output mod ~2^2048 would emit challenges in
+// [0, 2^256) instead of [0, N), giving the session-tagged path a strictly
+// weaker challenge distribution than the legacy HashToN path it shares the
```
Comment (pdyraga) [inline crypto/paillier/mod_proof.go]: Is the `HashToN` path legacy, or is it just a path without a session - a different option, valid for specific use cases? The comment might be misleading if the latter is true.

File: crypto/paillier/factor_proof_test.go [hunk 9]
```diff
@@ -117,6 +135,19 @@ func TestFactorProofVerifyFail3(t *testing.T) {
...
+	proof := privateKey.FactorProof(auxPrime.N, s, tt)
+	proof.Q = big.NewInt(0)
+	proof.Z1 = big.NewInt(-1)
+
+	assert.NotPanics(t, func() {
+		res, err := proof.FactorVerify(publicKey.N, auxPrime.N, s, tt)
+		assert.Error(t, err)
+		assert.False(t, res, "proof verify result must be false")
+	})
+}
```
Comment (pdyraga) [inline crypto/paillier/factor_proof_test.go:145]: Nice addition of the `FactorVerify` function. We could cover more cases than just the Q field and test them on non-zero values as well.

File: BNB_HARDENING_INTEGRATION.md
Comment (lukasz-zimnoch) [inline BNB_HARDENING_INTEGRATION.md:1]: FWIW, there is a new upstream PR that may be relevant: https://github.com/bnb-chain/tss-lib/pull/332

File: README.md
Comment (lukasz-zimnoch) [inline README.md:1]: Loose idea for follow-up work: Maybe a good move would be to remove all the code that is not used by `keep-core`, i.e. the whole `eddsa` and `ecdsa/resharing` packages? There are at least several issues in those packages that are inherited from upstream, for example:

1. **Resharing round 4 omits Paillier N and NTilde bit-length checks present in keygen, allowing weak-modulus key substitution** -> `ecdsa/keygen/round_2.go:53-61` rejects any peer-broadcast Paillier modulus or NTilde whose BitLen is not exactly paillierBitsLen (2048). The analogous loop in `ecdsa/resharing/round_4_new_step_2.go` performs DLN and mod-proof checks, h1!=h2, h1/h2 uniqueness — but NO BitLen validation of paiPK.N or NTildej. A malicious new-committee party can therefore broadcast a 1024-bit factorable N / a smooth-factor NTildej, generate valid DLN/Mod proofs over it (the proofs do not certify size), and have honest new-committee verifiers persist that weak key. This may open the door for a class of attacks that could lead to secret recovery given 1024-bit keys are considered vulnerable nowadays.
2. **EdDSA resharing has no SSID-bound proof transcript in this port** as the "Residual Risks" section of `BNB_HARDENING_INTEGRATION.md` mentions. 

... and probably more. Getting rid of those packages greatly limits the potential attack surface, and will make the `threshold-network/tss-lib` far easier to maintain. Moreover, we avoid dead code with known problems to bite us one day or bite another protocols that will leverage the affected packages of this fork.

File: BNB_HARDENING_INTEGRATION.md
Comment (mswilkison) [reply BNB_HARDENING_INTEGRATION.md:1]: Addressed in #4

File: README.md
Comment (mswilkison) [reply README.md:1]: Addressed in #5

File: crypto/mta/range_proof_test.go (hunk 4 above)
Comment (piotr-roslaniec) [reply crypto/mta/range_proof_test.go:177]: Fixed in 16d848b

File: crypto/paillier/mod_proof.go (hunk 8 above)
Comment (piotr-roslaniec) [reply crypto/paillier/mod_proof.go]: Fixed in https://github.com/threshold-network/tss-lib/pull/2/commits/630a7e57a5a1d29352c53ce86575bc51375792ce

**Review by piotr-roslaniec — APPROVED** [review]
